server/external/rapidapiClient.js

```diff
@@ -1,8 +1,78 @@
 const axios = require("axios").default;
+const { checkCache, getCachedCall, cacheData } = require('../redis.repository');
+
+
+const getFromCache = async (requestUrl) => {
+	try {
+        const dataString = await getCachedCall(requestUrl);
+        if (dataString !== null){
+            const data = JSON.parse(dataString);
+            return data;
+        }
+        else { 
+            console.log('expected data, but found none');
+            return null;
+        }
+	} catch (error) {
+	    console.log(error)
+	    return error
+	}
+}
+
+const pushToRedis = async (requestUrl, data) => {
+  // convert response object into JSON string, cache data in redis
+  const dataString = JSON.stringify(data);
+  return await cacheData(requestUrl, dataString);
+}
+
+
+/**
+ * This function completes a request by checking redis to see if cached response for API call exists.
+ * If it exists, this function returns the data as JSON. 
+ * If it does not exist, this function will make the API call, then cache the result in redis for future use.
+ * @param {String} requestUrl 
+ * @param {Object} options 
+ * @returns {Object} will always return the required data, either from cache or from API call
+ */
+const makeQuery = async (requestUrl, options) => {
+    // check for existence of key
+    try {
+        const keyExists = await checkCache(requestUrl);
+        if (keyExists === 1){
+            console.log('key exists in cache, checking now')
+            try {
+                const data = await getFromCache (requestUrl);
+                return data;
+            } catch (error) {
+                console.log(error);
+                return error;
+            }
+        } 
+        // data not in cache
+        else {
+            console.log('key not in cache, calling API');
+            try {
+                // make API call, then cache result for future use
+                console.log('making call')
+                const response = await axios.request(options);
+                const pushstatus = await pushToRedis(requestUrl, response.data);
+                console.log('data cache: ' + pushstatus);
+                return response.data;
+            } catch (error) {
+                console.log(error);
+                return error
+            }
+        }
+    } catch (error) {
+        console.log(error);
+        return error;
+    }
+}
+
 
 // this function makes an API call to the uNoGs unofficial netflix API, provided by rapidapi
 // changing the parameters of the API call is difficult, consult the rapidapi platform
-const getTop10 = () => {
+const getTop10 = async () => {
```

Comment:
I'm a bit confused with this method. When do we call it? Are we calling this method when nothing is in the cache?

Reply:
This method will always be called whenever the frontend requires a top10 list of movies. However, in this pull request, I've updated the logic inside this method to get the raw list of movies by calling the `makeQuery()` method, also in rapidapiClient.js. 

the `makeQuery()` method contains logic to first check the cache (redis), and then make an API call if nothing exists in the cache for this particular API query. Either way, the exact same data will be returned to the `getTop10()` method, which is only responsible for formatting the received data and sending it back to be emitted to the frontend. Let me know if that makes sense!

Reply:
That makes sense! Thanks for clearing it up.